### ISSUE TITLE
Object changed during "BeforeSave" is not being saved

### ISSUE THREAD

**Issue by wdepril** [jinzhu/gorm#859]
See below for observed behaviour for descr of the records missing.

```
package main

import (
	"fmt"

	_ "github.com/go-sql-driver/mysql"
	"github.com/jinzhu/gorm"
)

var db gorm.DB

type User struct {
	gorm.Model
	Name      string
	Locations []ObsLocation
}

type ObsLocation struct {
	gorm.Model
	UserId  uint
	Address ObsAddress
}

type ObsAddress struct {
	gorm.Model
	ObsLocationId uint
	Hash          string
	Street1       string
	NormAddrId    uint
	NormAddr      NormAddress
}

func (o *ObsAddress) BeforeSave() error {
	o.Hash = o.Street1

	var mss NormAddress
	err := db.Where(NormAddress{Hash: o.Hash}).First(&mss).Error

	if err != nil && err != gorm.RecordNotFound { //real error occured
		fmt.Printf("Error beforesave in Address %s\n", err.Error())
		return err
	} else if err == gorm.RecordNotFound { //normalized object does not exist yet
		fmt.Println("Record not found before save address")
		o.NormAddr.Hash = o.Hash
		o.NormAddr.Street = "test street"
	} else { // normalized object exists, initialize
		o.NormAddr = mss
	}

	return nil
}

type NormAddress struct {
	gorm.Model
	Hash   string
	Street string
	Post   string
	Nr     string
}

func init() {
	var err error
	db, err = gorm.Open("mysql", "root@/test?charset=utf8&parseTime=True&loc=Local")
	if err != nil {

		panic(err)
	}
	db.LogMode(true)
	db.SingularTable(true)

}

func main() {
	db.Exec("Drop table user")
	db.Exec("Drop table obs_location")
	db.Exec("Drop table obs_address")
	db.Exec("Drop table norm_address")

	if errs := db.AutoMigrate(
		&User{},
		&ObsLocation{},
		&ObsAddress{},
		&NormAddress{}).GetErrors(); errs != nil {
		fmt.Println("Error Migrating :", errs)
		return
	}

	var loc ObsLocation

	//Observed behaviour: NormAddr as filled in during the Before Save is not being added to the databse
	loc.Address = ObsAddress{Street1: "123 My Test Street"}
	u := User{Name: "depril", Locations: append(make([]ObsLocation, 0), loc)}

	if err := db.Create(&u).GetErrors(); err != nil {
		fmt.Println("Create Errors: ", err)
	}

	var loc2 ObsLocation

	//Observed Behaviour : NormAddr as seen below is not being added to the database
	loc2.Address = ObsAddress{Street1: "123 My Test Street",
		NormAddr: NormAddress{Street: "My Test",
			Post: "St",
			Nr:   "231"}}

	u2 := User{Name: "depril", Locations: append(make([]ObsLocation, 0), loc2)}

	if err := db.Create(&u2).GetErrors(); err != nil {
		fmt.Println("Create Errors: ", err)
	}
}
```

**Comment by jinzhu** [comment]
`NormAddress` has no valid relation with `ObsAddress` (no foreign key), so it won't be auto saved.

**Comment by wdepril** [comment]
Is there no automatic association created when ID are named the same as teh variable?
In Address there is :
NormAddrId    uint
NormAddr      NormAddress

**Comment by jinzhu** [comment]
Sorry, haven't notice that, then should be caused by some value is not addressable when saving in the callback, let me research that. 

Thank you for your report.

**Comment by jinzhu** [comment]
Hi @wdepril 

This issue should been fixed in `v1.0_dev` branch, refer: http://jinzhu.me/gorm/curd.html#change-updating-values-in-callbacks

You need to change your script, and add this line:
```go
scope.SetColumn("NormAddr", normAddr)
```

Full Example:

```go
package main

import (
	"fmt"

	_ "github.com/go-sql-driver/mysql"
	"github.com/jinzhu/gorm"
)

var db *gorm.DB

type User struct {
	gorm.Model
	Name      string
	Locations []ObsLocation
}

type ObsLocation struct {
	gorm.Model
	UserId  uint
	Address ObsAddress
}

type ObsAddress struct {
	gorm.Model
	ObsLocationId uint
	Hash          string
	Street1       string
	NormAddrId    uint
	NormAddr      NormAddress
}

func (o *ObsAddress) BeforeSave(scope *gorm.Scope) error {
	o.Hash = o.Street1

	var mss NormAddress
	err := db.Where(NormAddress{Hash: o.Hash}).First(&mss).Error

	if err != nil && err != gorm.RecordNotFound { //real error occured
		fmt.Printf("Error beforesave in Address %s\n", err.Error())
		return err
	} else if err == gorm.RecordNotFound { //normalized object does not exist yet
		fmt.Println("Record not found before save address")
		normAddr := o.NormAddr
		normAddr.Hash = o.Hash
		normAddr.Street = "test street"
		scope.SetColumn("NormAddr", normAddr)
		fmt.Println(normAddr)
	} else { // normalized object exists, initialize
		o.NormAddr = mss
	}

	return nil
}

type NormAddress struct {
	gorm.Model
	Hash   string
	Street string
	Post   string
	Nr     string
}

func init() {
	var err error
	db, err = gorm.Open("mysql", "gorm:gorm@/gorm?charset=utf8&parseTime=True&loc=Local")
	if err != nil {

		panic(err)
	}
	// db.LogMode(true)
	db.SingularTable(true)

}

func main() {
	db.Exec("Drop table user")
	db.Exec("Drop table obs_location")
	db.Exec("Drop table obs_address")
	db.Exec("Drop table norm_address")

	if errs := db.AutoMigrate(
		&User{},
		&ObsLocation{},
		&ObsAddress{},
		&NormAddress{}).GetErrors(); errs != nil {
		fmt.Println("Error Migrating :", errs)
		return
	}

	var loc ObsLocation

	//Observed behaviour: NormAddr as filled in during the Before Save is not being added to the databse
	loc.Address = ObsAddress{Street1: "123 My Test Street"}
	u := User{Name: "depril", Locations: append(make([]ObsLocation, 0), loc)}

	if err := db.Debug().Create(&u).GetErrors(); err != nil {
		fmt.Println("Create Errors: ", err)
	}

	var loc2 ObsLocation

	//Observed Behaviour : NormAddr as seen below is not being added to the database
	loc2.Address = ObsAddress{Street1: "123 My Test Street",
		NormAddr: NormAddress{Street: "My Test",
			Post: "St",
			Nr:   "231"}}

	u = User{Name: "depril", Locations: append(make([]ObsLocation, 0), loc)}

	if err := db.Create(&u).GetErrors(); err != nil {
		fmt.Println("Create Errors: ", err)
	}
}
```

Thank you for your feedback

**Comment by stroborobo** [comment]
@jinzhu So is this only for struct members? In your example you didn't use `scope.SetColumn` for the `o.Hash = o.Street1`. I'm a little confused since in the guide you're talking about values.
